asgi_ipc.py

```diff
@@ -189,188 +164,113 @@ def flush(self):
-        self.mmap.seek(0)
-        self.mmap.write(b"\0\0\0\0\0\0\0\0")
         # Unlink and remake the semaphore
         self.semaphore.unlink()
         self.semaphore = posix_ipc.Semaphore(
-            self.path + "-sem",
+            self.identifier + "-sem",
             flags=posix_ipc.O_CREX,
             mode=0o660,
             initial_value=1,
         )
 
-    def __del__(self):
-        """
-        Explicitly closes the shared memory area.
-
-        The semaphore is not closed as this is not threadsafe; closing it
-        prevents any other layers in the same process from seeing it.
-        """
-        self.mmap.close()
-        self.shm.close_fd()
-        self.shm.unlink()
-
     def flush(self):
-        self._set_value(self.datatype())
+        self.connection.cursor().execute('DELETE FROM {table_name}'.format(table_name=self.table_name))
+        self.connection.commit()
 
-    def __iter__(self):
-        return iter(self._get_value())
+    def _execute(self, query, *args):
+        try:
+            self.semaphore.acquire(self.death_timeout)
+        except posix_ipc.BusyError:
+            self._reset()
+            self.semaphore.acquire(0)
+        try:
+            cursor = self.connection.cursor()
+            cursor.execute(query.format(table=self.table_name), args)
+            result = cursor.fetchall()
+        finally:
+            self.semaphore.release()
+        return result
 
-    def __contains__(self, item):
-        return item in self._get_value()
 
-    def __getitem__(self, key):
-        return self._get_value()[key]
+class MessageTable(SqliteTable):
+    table_name = 'messages'
+    table_structure = '''
+        CREATE TABLE IF NOT EXISTS messages
+        (channel text, message text, expiry datetime)
+    '''
 
-    def __setitem__(self, key, value):
-        d = self._get_value()
-        d[key] = value
-        self._set_value(d)
+    def get_messages(self, channel) -> list:
+        return self._execute('SELECT message, expiry FROM {table} WHERE channel=?', channel) or (None, None)
 
-    def __delitem__(self, key):
-        d = self._get_value()
-        del d[key]
-        self._set_value(d)
+    def get_message_count(self, channel) -> int:
+        return self._execute('SELECT COUNT(*) FROM {table} WHERE channel=?', channel)[0][0]
 
-    def __len__(self):
-        return len(self._get_value())
+    def add_message(self, message, expiry, channel):
+        self._execute('INSERT INTO {table} (channel, message, expiry) VALUES (?,?,?)', channel, message, expiry)
 
+    def pop_message(self, channel) -> tuple:
+        result = self._execute('SELECT message, expiry FROM {table} WHERE channel=? LIMIT 1', channel)
+        if not result:
+            raise ValueError('No message in channel')
+        result = result[0]
+        self._execute('DELETE FROM {table} WHERE channel=? AND message=? AND expiry=?', channel, result[0], result[1])
```

Comment:
Though, thinking about it, you'd need to have the exact same message sent at the exact same time. But that is possible, and it would probably make the delete shorter too.